beacon_node/eth2-libp2p/src/peer_manager/mod.rs

```diff
@@ -261,7 +271,91 @@ impl<TSpec: EthSpec> PeerManager<TSpec> {
+        debug!(self.log, "RPCError"; "protocol" => protocol.to_string(), "err" => err.to_string());
+
+        // Map this error to a `PeerAction` (if any), and decide if the peer manager should request
+        // some `PeerManagerEvent`
+        let (peer_action, event) = match err {
+            RPCError::IncompleteStream => {
+                // TODO: implement
+                (None, None)
+            }
+            RPCError::InternalError(_reason) => {
+                // Our fault. Now they hate us. Do nothing
+                (None, None)
+            }
+            RPCError::InvalidData => {
+                // Peer is not complying with the protocol. This is considered a malicious action
+                (
+                    Some(PeerAction::Malicious),
+                    Some(PeerManagerEvent::BanPeer(peer_id.clone())),
+                )
+            }
+            RPCError::IoError(_e) => {
+                // no reason to consider an IoError as a Malicious action
+                (Some(PeerAction::HonestError), None)
+            }
+            RPCError::ErrorResponse(code) => {
+                // TODO verify this logic
+                match code {
+                    RPCResponseErrorCode::ServerError => (None, None),
+                    RPCResponseErrorCode::InvalidRequest => (None, None),
+                    RPCResponseErrorCode::Unknown => (None, None),
+                }
+            }
+            RPCError::SSZDecodeError(_) => (Some(PeerAction::Malicious), None),
+            RPCError::UnsupportedProtocol => {
+                // Not supporting a protocol shouldn't be considered a malicious action, but
+                // it is an action that in some cases will make the peer unfit to continue
+                // communicating.
+                // TODO: To avoid punishing a peer repeatedly for not supporting a protocol, this
+                // information should be stored and used to prevent sending requests for the given
+                // protocol to this peer. Similarly, to avoid blacklisting a peer for a protocol
+                // forever, this information should expire.
+                let action = Some(PeerAction::HonestError);
+
+                // TODO: verify this logic
+                let event = match protocol {
+                    Protocol::Ping => Some(PeerManagerEvent::DisconnectPeer(peer_id.clone())),
+                    Protocol::BlocksByRange => None,
+                    Protocol::BlocksByRoot => None,
+                    Protocol::Goodbye => None,
+                    Protocol::MetaData => Some(PeerManagerEvent::DisconnectPeer(peer_id.clone())),
+                    Protocol::Status => Some(PeerManagerEvent::DisconnectPeer(peer_id.clone())),
+                };
+
+                (action, event)
+            }
+            RPCError::StreamTimeout => {
+                let action = Some(PeerAction::TimedOut);
+                // TODO: verify this logic
```

Comment:
I think potentially we tolerate 2 stream timeouts for the others. Then we disconnect them. 
Ping happens every 30 seconds, so we might want to disconnect earlier if the others are all timing out